execute/profiler_test.go

```diff
@@ -18,50 +18,136 @@ import (
+		op := fmt.Sprintf("type%d", i%2)
+		go fn(op, "lab0", ctx, 100+i*i)
+		// Waiting between each loop seems to ensure better consistency of row order
+		// for final table. However, because the result aggregates are stored in a map,
+		// the order in which the rows are written cannot be guaranteed
+		time.Sleep(100 * time.Millisecond)
 	}
+	wantStr.WriteString(fmt.Sprintf(",,0,profiler/operator,%s,%s,%d,%d,%d,%d,%f\n",
+		"type0",
+		"lab0",
+		2,
+		1100,
+		1104,
+		2204,
+		1102.0,
+	))
+	wantStr.WriteString(fmt.Sprintf(",,0,profiler/operator,%s,%s,%d,%d,%d,%d,%f\n",
+		"type1",
+		"lab0",
+		2,
+		1101,
+		1109,
+		2210,
+		1105.0,
+	))
+	wg.Wait()
+	// Wait a bit for the profiling results to be added.
+	// In the query code path this is guaranteed because we only access the result
+	// after the query finishes execution AND its result tables are read and encoded.
+	time.Sleep(100 * time.Millisecond)
+	tbl, err := p.GetResult(nil, &memory.Allocator{})
+	if err != nil {
+		t.Error(err)
+	}
+	result := table.NewProfilerResult(tbl)
+	got := flux.NewSliceResultIterator([]flux.Result{&result})
+	dec := csv.NewMultiResultDecoder(csv.ResultDecoderConfig{})
+	want, e := dec.Decode(ioutil.NopCloser(strings.NewReader(wantStr.String())))
+	if e != nil {
+		t.Error(err)
+	}
+	if err := executetest.EqualResultIterators(want, got); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestOperatorProfiler_GroupByLabel(t *testing.T) {
+	// Create an operator profiler
+	p := execute.AllProfilers["operator"]()
+	// And inject it to the context.
+	ctx := context.WithValue(context.Background(), execute.OperatorProfilerContextKey, p)
+	// Build the "want" table.
+	var wantStr bytes.Buffer
+	wantStr.WriteString(`
+#datatype,string,long,string,string,string,long,long,long,long,double
+#group,false,false,true,false,false,false,false,false,false,false
+#default,_profiler,,,,,,,,,
+,result,table,_measurement,Type,Label,Count,MinDuration,MaxDuration,DurationSum,MeanDuration
+`)
+	count := 4
+	wg := sync.WaitGroup{}
+	wg.Add(count)
+	fn := func(opType string, label string, ctx context.Context, offset int) {
+		st := time.Date(2020, 10, 14, 12, 30, 0, 0, time.UTC)
+		_, span := execute.StartSpanFromContext(ctx, opType, label, opentracing.StartTime(st))
+		profilerSpan := span.(*execute.OperatorProfilingSpan)
+		// Finish() will write the data to the profiler
+		// In Flux runtime, this is called when an execution node finishes execution
+		profilerSpan.FinishWithOptions(opentracing.FinishOptions{
+			FinishTime: time.Date(2020, 10, 14, 12, 30, 0, 1000+offset, time.UTC),
+		})
+		wg.Done()
+	}
+	for i := 0; i < count; i++ {
+		label := fmt.Sprintf("lab%d", i%2)
+		go fn("type0", label, ctx, 100+i*i)
+		// Waiting between each loop seems to ensure better consistency of row order
+		// for final table. However, this is still not guaranteed, because the result
+		// aggregates are stored in a map, and thus order is not 100% guaranteed.
+		time.Sleep(100 * time.Millisecond)
```

Comment:
Wanted to draw attention to this. It seems that this was a problem in the old version of the test because of concurrency issues. Now it's a problem because of concurrency issues *and* the fact that we can't guarantee the order of values in a map.

This shouldn't be a problem in production, but it will cause these two tests to fail every once in a while. Not sure if there's a good way around that.

Reply:
This problem has been resolved by: https://github.com/influxdata/flux/pull/3262/commits/3c991325cbb8e015c8561bfc7ebddaff19cd06be